### PR TITLE
Run MOVE_COMPILER_V2 with more tests and record outputs in .v2_exp files

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Updated various test harnesses and re-ran pretty much all .move tests with `MOVE_COMPILER_V2=true` to generate .v2_exp files to see what V2 does today.  Diffed outputs with V1 outputs and filed bugs for differences (fortunately few).

Tests were run manually.  Perhaps there should be a script.  Something like the following was used:
```
for value in false true; do 
UB=1 MOVE_COMPILER_V2=$value cargo test \
      -p "aptos-transactional-test-harness" \
      -p "bytecode-verifier-transactional-tests" \
      -p "move-async-vm" \
      -p "move-cli" \
      -p "move-model" \
      -p "move-package" \
      -p "move-prover-bytecode-pipeline" \
      -p "move-stackless-bytecode" \
      -p "move-to-yul" \
      -p "move-transactional-test-runner" \
      -p "move-unit-test"
      -p "move-vm-transactional-tests" \
      --no-fail-fast \
      >& test-$value.out
done
```

Note, did not yet set `MOVE_COMPILER_BLOCK_V1` so some of these may be using V1 under the hood, but with the baseline outputs checked in we can see if things degrade when we force use of V2.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
See above.

### CODE REVIEW

File: third_party/move/evm/move-to-yul/tests/AccountStateMachine.v2_exp [hunk 1]
```diff
@@ -0,0 +1,1138 @@
+/* =======================================
```
Comment (wrwg) [inline third_party/move/evm/move-to-yul/tests/AccountStateMachine.v2_exp:1]: I think I said before we should keep the evm stuff alone. Why are we adding this additional complexity?

File: third_party/move/extensions/async/move-async-vm/tests/sources/AccountStateMachine.v2_exp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+publishing Actor
```
Comment (wrwg) [inline third_party/move/extensions/async/move-async-vm/tests/sources/AccountStateMachine.v2_exp:1]: Same here. Once we switch to v2 compiler as the default, those things can be updated. This does not give us relevant new test coverage, most test cases here and in the EVM code are not specific to Move features, but to the tool.

File: third_party/move/move-bytecode-verifier/transactional-tests/tests/ability_field_requirements/fields_conditional_abilities.v2_exp [hunk 3]
```diff
@@ -0,0 +1,37 @@
+processed 8 tasks
```
Comment (wrwg) [inline third_party/move/move-bytecode-verifier/transactional-tests/tests/ability_field_requirements/fields_conditional_abilities.v2_exp:1]: How does that differ from the simple .exp? There shouldn't be any difference. Is it the definition indices?

**Review by wrwg — COMMENTED** [review]
- I'm not seeing what this PR brings us in new quality of testing since most features we have v2 tests for now are rather generic Move code which very likely are already covered by our regular tests.
- Instead we are adding new technical debt by having the v1/v2 living side by side. When we remove v1, we have to pay for this.
- Some of the v2 tests (e.g. bytecode verifier) should have the same .exp file and not a new one. If the difference is in things like definition index we should better redact this from the baseline instead of creating new v2_exp.

File: third_party/move/move-command-line-common/src/testing.rs [hunk 4]
```diff
@@ -11,6 +11,14 @@ pub const EXP_EXT: &str = "exp";
 /// Extension for expected output files compiled by v2
 pub const EXP_EXT_V2: &str = "v2_exp";
 
+pub fn get_compiler_exp_extension() -> &'static str {
+    if get_move_compiler_v2_from_env() {
```
Comment (wrwg) [inline third_party/move/move-command-line-common/src/testing.rs:15]: Any tests which are dependent from env variables must be run in their own CI action and cannot run in parallel with other tests and different env vars. Its perhaps OK we do this for e2e tests (like we do already) but for unit tests it is not a good design. Where we have already different exp files (prover, compiler e2e tests) we do not do this via env vars but via test configs.

**Comment by brmataptos** [comment]
> * I'm not seeing what this PR brings us in new quality of testing since most features we have v2 tests for now are rather generic Move code which very likely are already covered by our regular tests.

I've already been able to somewhat reproduce Kevin's bug using move-docgen tests, because of the different build environment provided.  In other cases, the added coverage is small, but it is non-zero.  We can also test which tools/paths accidentally still use V1 under the covers by using the MOVE_COMPILER_DISABLE_V1 env var.

> * Instead we are adding new technical debt by having the v1/v2 living side by side. When we remove v1, we have to pay for this.
> * Some of the v2 tests (e.g. bytecode verifier) should have the same .exp file and not a new one. If the difference is in things like definition index we should better redact this from the baseline instead of creating new v2_exp.

This gives us an easy way to get a bit more coverage to track compiler V2 correctness and avoiding regressions:

- Check in this PR, after I 
   - (1) carefully examined diffs between corresponding .exp and .v2_exp files, and
   - (2) filed bugs for any meaningful differences.
- Occasionally or regularly (CI) run more tests with MOVE_COMPILER_V2 to see any changes and fix them.
- Eventually also run with MOVE_COMPILER_DISABLE_V1 to verify that V1 is truly unused.

A few parallel output files which we can ignore (after merging this PR) unless they change is an insignificant debt to take on for the benefit, even if it is small.  Trying more complex approaches to leverage the testing value of these other packages would involve a significant amount of wasted time or technical debt: Jumping through hoops to make outputs identical, or building complex redaction/comparison frameworks to try to enable V2 runs to use exactly the same .exp files as V1 is technical debt.  In retrospect, the move-compiler-v2-transactional-tests comparison mechanism might have been better done this way: we would avoided wasteful duplicate runs of V1 on every test, and also simplified the testing code significantly.

The cost is a one-time comparison of each .v2_exp file with each .exp file, and then we can forget about v1.

We can rename .v2_exp to .exp when we get rid of V1 if it bothers you.

**Review by wrwg — APPROVED** [review]
Alright, lets land it.

**Comment by rahxephon89** [comment]
Hi @brmataptos, sorry I don't look all code and new files in detail because there are so many. I am wondering how these tests are run against the compiler V2 on CI? Do we automatically run V1 and then V2 for all of them (like the prover test)?

**Comment by brmataptos** [comment]
> Hi @brmataptos, sorry I don't look all code and new files in detail because there are so many. I am wondering how these tests are run against the compiler V2 on CI? Do we automatically run V1 and then V2 for all of them (like the prover test)?

I think these will have to be run as a separate CI job with `MOVE_COMPILER_V2=true` set in the environment, as (1) Wolfgang has observed that varying environment variables across tests causes races, and (2) these tests represent a collection of tools, and feeding a CLI flag to V2 in each case will be tricky.  While the move-prover example isolates the multiple configurations better, it's a lot of work for each test framework to implement.

The problem is that these tests represent a number of different test frameworks, so doing something like what was done for the prover tests would require a lot of work of questionable long-term value.  This is a quick hack that allows us to run tests in the following directories under V2 with minimal effort, gaining perhaps 637 files with very few changes to .rs files.  It requires some scripting to manage this collection of tests, but it will be easy to rename and/or change the defaults once V2 becomes mainstream.

Note that this PR introduces test outputs for the following directories:
```
aptos-move/aptos-transactional-test-harness/tests/
third_party/move/evm/move-to-yul/tests/
third_party/move/extensions/async/move-async-vm/tests/
third_party/move/move-bytecode-verifier/transactional-tests/tests/
third_party/move/move-model/bytecode/tests/
third_party/move/move-model/tests/
third_party/move/move-prover/bytecode-pipeline/tests/
third_party/move/move-vm/transactional-tests/tests/
third_party/move/testing-infra/transactional-test-runner/tests/
third_party/move/tools/move-cli/tests/
third_party/move/tools/move-package/tests/
third_party/move/tools/move-unit-test/tests/
```

Of the total 835 files '*.v2_exp' I see, just 110 of them have any difference from the `.exp` files, with a selective diff output of 1667 lines.
[diff-out.txt](https://github.com/aptos-labs/aptos-core/files/15367535/diff-out.txt)

Of these files:
- many tests have trivial output differences (such as message wording, or different redactions in prover output)
- the aptos_test_harness files have some differences, partially addressed in PR #13337; after one of these PRs lands I can file more targeted bugs from seeing the other.
- ./third_party/move/testing-infra/transactional-test-runner/tests/vm_test_harness/example.v2_exp:
    - shows that V2 disallows a module publication update
    - filed issue #13255 
- ./third_party/move/tools/move-cli/tests/build_tests/disassemble_script/args.v2_exp shows:
        - V2 doesn't do very well at optimizing the code
        - other move-cli tests show that V2's code modules are bigger than V1: 
            - added examples to #12404 for further study
-  V2 rejects two tests because packages have different address aliases; this was allowed in old Move but not Aptos 
    - ./third_party/move/tools/move-package/tests/test_sources/compilation/multiple_deps_rename/Move.v2_exp
    - ./third_party/move/tools/move-package/tests/test_sources/compilation/multiple_deps_rename_one/Move.v2_exp
    - Filed Issue #13256

File: third_party/move/evm/move-to-yul/tests/AccountStateMachine.v2_exp (hunk 1 above)
Comment (brmataptos) [reply third_party/move/evm/move-to-yul/tests/AccountStateMachine.v2_exp:1]: It was easy enough to add it.  Just a line in a script.

**Review by rahxephon89 — APPROVED** [review]
I am wondering whether we need to add a script to this PR so that after any changes to the compiler, we can use it to automatically update these exp files?

**Pull request merged**